python3-diff.md

```diff
@@ -0,0 +1,42 @@
+# Differences between Python 3 and Starlark
+
+Starlark language is very similar to Python. Starlark tries to follow Python
+syntax and semantics where possible, but full compatibility with Python is neither possible nor a goal.
+
+This document describes some differences between Python 3 and Starlark.
+The list is incomplete.
+
+This document is not normative.
+
+## The differences
+
+* Starlark evaluation is deterministic and should produce the identical output in any implementation
+
+### Language
```

Comment:
I would separate this into "syntax", "semantics", and "core library".

The true of list of differences is much, much longer than what we have here, too long to enumerate in full, so I wonder what should be our rationale for inclusion. Perhaps we could summarize each category in a single item, e.g.

- Starlark does not support classes, exceptions, generators [etc].
- Starlark uses load statements in place of imports. Modules are not (currently) first class.
- Starlark data types provide only a subset of the corresponding Python ones, notably excluding "double-underscore" methods such as `__add__`, `__mul__` etc.
- Starlark's library or core function is a subset of Python's.

Reply:
Cribbing from https://docs.bazel.build/versions/master/skylark/language.html#differences-with-python, which should be made to link here:
- variables created by a thread (e.g. when initializing a module) become frozen
- no rebinding of globals (though impls may have a flag)
- no recursion nor while loops (though impls may have a flag)
- no for/if at top level (though impls may have a flag)
- dict iterators use insertion order
- ordered comparison < requires operands of same type (or int < float)
- float NaNs are totally ordered
- no trailing commas in unparenthesized tuples
- no dup keys in dict expressions (irritatingly)
- strings are not iterable
- no implicit string concatenation (use +)
- no set type (Go impl has a flag)
- no `is` operator (even though the language does have mutation)
- no assignments to lexically enclosing variables

Reply:
> what should be our rationale for inclusion

To get a quick answer: how this particular thing works in Starlark.

For Python it is easy: just look how it is implemented in CPython, which is a reference implementation.

Bazel Starlark is not a reference implementation: it does not follow the spec precisely and changes often.

Spec should be the reference, but:
* sometimes the spec is [underspecified](https://github.com/bazelbuild/starlark/pull/142) or [buggy](https://github.com/bazelbuild/starlark/pull/155)
* spec is somewhat hard to read

So in 99% of cases, to answer "how this particular thing works in Starlark" it could be enough to look at Python plus this document.

Could be useful for people:
* working on Starlark implementations
* or tooling
* of simply writing some Starlark code

And another reason, perhaps the most important one. If there's a difference between Python and Starlark, how get an answer is that intentional or mistake? The answer would be:
* if it is listed in the document, the difference is intentional
* otherwise it is not, and something need to be fixed (either spec or this document)

Reply:
> Starlark's library of core functions is a subset of Python's.

It is not a subset:
* strings have `elems` which is not in Python
* certain functions like `type` have different semantics

Reply:
> dict iterators use insertion order

Python does it too since 3.7.

Reply:
> ordered comparison < requires operands of same type (or int < float)

Same as in Python?

```
>>> "1" < []
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'str' and 'list'
```

Reply:
True. I was thinking of Python2, but I suppose we've staked v3 as the baseline for comparison.